### PR TITLE
Improve Language Properties (Language/Language Native/Variant/Variant Native)

### PR DESCRIPTION
**Pull request by julkue**
I've created this PR, as I've noticed that I've done a mistake in the `native` property inside ch.json. I thought, it's the native property for the `variant`, but it's the native property of `language`. Then I found out, that we don't have a correct property for the language variant written in the native language (in ch.json it's `Schweizerdeutsch`). We have a  `language` meta key for the language written in English, and a `native` meta key for the language written in the native language. But we only have one `variant` key, for the variant written in English.  I think we also need a native variant key, for the variant written in the native language.

Also, I'm not sure how we should handle the case sensity in the decompose value. An `Ä` can decompose into `Ae` if the word is `Äpfel`, but it should decompose into `AE` if the word is `ÄPFEL`...

Also I've noticed that a [country property](https://github.com/diacritics/database/blob/df7a44080100d7a534b572cef3e634390062d4f8/v1/diacritics.json#L171) is automatically generated for the German language variant, but not for Swiss Standard German.

@Mottie What do you think?

### CODE REVIEW

**Comment by Mottie** [comment]
Sorry, I've been so busy that I neglected to follow up on this PR. I'm not sure why #19 was added along with #17, but I think we either need to address the definitions of "language" and "variant" in the spec or add an optional "native-variant". I think if we go with updating the "language" and "variant", we may also need to add a "parent-language" entry. With a "native-variant" addition, it would be optional and only be added for variant languages.

As for the casing of the decomposed value, the database itself doesn't get access to the word, so the addition would need to be made in the node module. It may be necessary to update the spec for "decompose" to include an object:

```js
"Ä": {
  "mapping": {
    "base": "A",
    "decompose": {
      "title": "Ae",
      "caps": "AE"
    }
  }
}
```

**Comment by julkue** [comment]
> or add an optional "native-variant"

I'd vote for this one:

- `language`
- `language-native`
- `variant`
- `variant-native`

> we may also need to add a "parent-language" entry

For which purpose?

> so the addition would need to be made in the node module.

Yeah, respectively in third party applications using a RegExp.

> It may be necessary to update the spec for "decompose" to include an object:

This sounds like a good idea. But two questions:

- Would this be possible at all, since we have genereated most of the language files automatically?
- What's the meaning behind the name `caps`? I mean I know that this would referring the scenario when the diacritic occurs in a normal sentence, but why "caps"?

We'd then need to ask @hadithmv again for updating the JSON... Hopefully he's open for that.

**Comment by Mottie** [comment]
>> we may also need to add a "parent-language" entry
>
> For which purpose?

That was the alternative, if we didn't want to add `variant-native`.

> Would this be possible at all, since we have genereated most of the language files automatically?

Well, for the unverified files, we could programmatically update the `decompose` entry and use `toUpperCase()` on the desired characters.

> What's the meaning behind the name caps?

It's an abbreviation for capitals... I was thinking "all-caps", but since I was in a hurry this morning and I wasn't sure what your thoughts were on the subject, I simply typed in "caps". I think "title-case" and "all-capitals" would be the long name versions of the entries; but do we want to use hyphens, or camel-cased keys?

**Comment by julkue** [comment]
> That was the alternative, if we didn't want to add variant-native.

I'd vote for `variant-native` as I personally find `parent` confusing.

> "title-case" and "all-capitals" 

That sounds easier to understand.

> do we want to use hyphens, or camel-cased keys?

I'd vote for camelCase, as it's easier to access in JS. For example `language.language-variant` isn't valid, `language.languageVariant` is.

> use toUpperCase() on the desired characters.

I'll try to update the schema and the German language files. Can you update the generator to use `toUpperCase()`?  
I've just noticed that the es.json has e.g. no `ue` mapping für `ü`. While the provided source in es.json doesn't include the `ue` mapping, it also doesn't inlcude in for Swiss German (btw.: Note the name, instead of Swiss Standard German. I'll rename it to this too, as I don't see a reason for the "standard"). Where did you find that there's no `ue` mapping?

**Comment by julkue** [comment]
Oh, and could please also take a look at this:

> Also I've noticed that a [country property](https://github.com/diacritics/database/blob/df7a44080100d7a534b572cef3e634390062d4f8/v1/diacritics.json#L171) is automatically generated for the German language variant, but not for Swiss Standard German.

**Comment by Mottie** [comment]
I may not be able to look at this until tomorrow...

https://en.wikipedia.org/wiki/Swiss_Standard_German#Written_Swiss_Standard_German

> The Swiss keyboard layout has no <kbd>ß</kbd> key, nor does it have the capital umlaut keys Ä, Ö and Ü. This dates back to mechanical typewriters that had the French diacritical marks letters on these keys to allow the Swiss to write French on a Swiss German QWERTZ keyboard (and vice versa). Thus a Swiss German VSM keyboard has an <kbd>ä</kbd> key that prints an à (a-grave) when shifted.[11] However, it is possible to write uppercase umlauts by use of caps lock or by using the <kbd>¨</kbd> dead key.

**Comment by julkue** [comment]
@Mottie Alright, thanks.

**Comment by Mottie** [comment]
Hmm,  ok so the CLDR database stores the territories under the ISO 639-3 language code for Swiss German which is "gsw" - [ref](https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/languageData.json#L2144-L2157). So we would either need to include that language code in the metadata or build a cross-reference.

**Comment by julkue** [comment]
So, from my point of view the automatic generation doesn't work as the database which is necessary for the generation uses ISO 639-2, while our database uses ISO 639-1. 

From my short research it seems that ISO 639-1 doesn't include all language codes we need, as Swiss German for example is only covered in ISO 639-2. However, I personally think that `de` will lead to a better usability than `deu`.

What's the best solution in your opinion?

**Comment by Mottie** [comment]
Well, I think we should drop ISO 639-1 from the spec completely since it is missing many language variants. As we are already using [IETF language tags](https://datahub.io/core/language-codes#resource-ietf-language-tags) for the variants, I think that's the easiest choice for an alternative. It uses ISO 639-1 as a base and includes the `-{variant}` as needed.

For this issue, I think adding a cross-reference between IETF and ISO 639-2 would be the best choice - I've been searching for one inside the CLDR data, but I haven't been successful yet. This change won't require us to change the spec (much), other than switching to IETF. The current file names would all remain the same.

**Comment by julkue** [comment]
Okay. Honestly, I haven't studied IETF and ISO 369, but what you say seems to make sense. 

> This change won't require us to change the spec (much)

What would be necessary to change at all?

>I've been searching for one inside the CLDR data, but I haven't been successful yet

Maybe we could parse this file?

http://loc.gov/standards/iso639-2/ISO-639-2_utf-8.txt

**Comment by Mottie** [comment]
>> This change won't require us to change the spec (much)
>
>What would be necessary to change at all?

Just change the spec references from ISO-639 to IETF, I think

> I think adding a cross-reference between IETF and ISO 639-2

I totally messed that up... we don't need a cross-reference. I'll dig into the language data a bit to update the properties in the list.

File: spec/README.md [hunk 1]
```diff
@@ -160,19 +161,27 @@ Type: `String`
...
 
 Optional  
 Type: `String`
 
 The associated language variant written in English, if applicable. For example, if the [IETF language tag](https://www.w3.org/International/articles/language-tags/) was `de_AT`, this entry would include the full name of the variant country, `Austria` for the `AT` variant.
 
-###### metadata.native
+###### metadata.variantNative
 
 Required
```
Comment (Mottie) [inline spec/README.md]: Shouldn't this be "Optional" since the `metadata.variant` is optional?

Comment (Mottie) [inline spec/README.md]: Shouldn't this be "Optional" since the `metadata.variant` is optional?

**Review by Mottie — CHANGES_REQUESTED** [review]
Everything else looks good!

File: src/de/ch.json [hunk 2]
```diff
@@ -5,8 +5,9 @@
       "EU"
     ],
     "language": "German",
-    "native": "Schweizerdeutsch",
-    "variant": "Swiss Standard German",
+    "languageNative": "Deutsch",
+    "variant": "Swiss German",
```
Comment (Mottie) [inline src/de/ch.json]: Actually, according to [Wikipedia](https://en.wikipedia.org/wiki/Swiss_German)

> The dialects of Swiss German must not be confused with Swiss Standard German, the variety of Standard German used in Switzerland.

I don't think this should be changed here.... IETF `de-CH` is *not* the same as ISO-639-2 `gsw`.

File: spec/README.md (hunk 1 above)
Comment (julkue) [reply spec/README.md]: My fault, fixed.

**Comment by julkue** [comment]
> Just change the spec references from ISO-639 to IETF, I think

Quote from the docs:

> The variant file will be named according to IETF language tag extended language (extlang) subtag. So, if a de_AT would be necessary, then the file would be named at.js; by the extended language subtag and in all lower-case.

Could you please elaborate what needs to be changed? From my point of view the variant already uses IETF.

> I totally messed that up... we don't need a cross-reference. I'll dig into the language data a bit to update the properties in the list.

Cool, thanks.

**Review by Mottie — COMMENTED** [review]
> Could you please elaborate what needs to be changed?

You're right, I looked through the spec and there aren't any changes to the ISO 639 references.

File: src/de/ch.json (hunk 2 above)
Comment (julkue) [reply src/de/ch.json]: > I don't think this should be changed here....

Actually it doesn't matter. Diacritics are the same in both `Swiss German` and `Swiss Standard German`. The latter is the name for when you write, the other when you speak. Anyway, changed.

**Review by Mottie — APPROVED** [review]

**Comment by hadithmv** [comment]
A bit late, but I see that I've been tagged.

Is there anything that needs doing? @julmot  @Mottie

**Pull request merged**